Library/Homebrew/extend/os/linux/install.rb

```diff
@@ -76,23 +76,43 @@ def symlink_ld_so
     end
     private_class_method :symlink_ld_so
 
-    def symlink_gcc_libs
+    def setup_preferred_gcc_libs
       gcc_opt_prefix = HOMEBREW_PREFIX/"opt/#{OS::LINUX_PREFERRED_GCC_FORMULA}"
+      ld_so = HOMEBREW_PREFIX/"opt/glibc/lib/ld-linux-x86-64.so.2"
+
+      if ld_so.readable?
+        ld_so_conf_d = HOMEBREW_PREFIX/"etc/ld.so.conf.d"
+        ld_so_conf_d.mkpath
+        FileUtils.chmod "go-w", ld_so_conf_d
+
+        # Add gcc to ld search paths
+        ld_gcc_conf = ld_so_conf_d/"50-homebrew-preferred-gcc.conf"
+        ld_gcc_conf.atomic_write <<~EOS
+          #{gcc_opt_prefix}/lib/gcc/#{GCC_VERSION_SUFFIX}
+        EOS
+        FileUtils.chmod "u=rw,go-wx", ld_gcc_conf
+      end
 
       GCC_RUNTIME_LIBS.each do |library|
-        gcc_library = gcc_opt_prefix/"lib/gcc/#{GCC_VERSION_SUFFIX}/#{library}"
         gcc_library_symlink = HOMEBREW_PREFIX/"lib/#{library}"
-        # Skip if the link target doesn't exist.
-        next unless gcc_library.readable?
 
-        # Also skip if the symlink already exists.
-        next if gcc_library_symlink.readable? && (gcc_library_symlink.readlink == gcc_library)
+        if ld_so.readable?
+          # Remove legacy symlinks
+          FileUtils.rm gcc_library_symlink if gcc_library_symlink.symlink?
+        else
```

Comment:
Given that `glibc` will be keg-only:

I wonder if we should just keep using our `glibc` even if the host's is newer.

This will allow us to get rid of these GCC library symlinks entirely.

Reply:
> I wonder if we should just keep using our `glibc` even if the host's is newer.

Are you saying 1) "we should use it even if the host's is newer" or 2) "we should not use it even if the host's is newer"?

Personally, I'm in strongly in favour of 2) and moderately strongly opposed to 1).

Reply:
It's actually neither; apologies for being unclear.

This `else` bit is the only part that keeps these GCC library symlinks around. This handles the case where the host Glibc is newer than ours, but the host GCC is older than our preferred version.

I was suggesting that for this case we should consider using a keg-only Glibc. I think that should allow us to get rid of these GCC symlinks entirely, which, based on previous discussion, was something we wanted to do but decided to punt on.

Whenever the host Glibc and host GCC are both not older, we should definitely not install the `glibc` formula.

Reply:
This is a really interesting idea although @sjackman should comment when he is back on it as well.  Using `ldconfig` as we are proposing to do would let us get rid of the GCC symlinks and hopefully move us closer towards getting rid of  any use of `-L#{HOMEBREW_PREFIX/}lib` and `-Wl,-rpath,#{HOMEBREW_PREFIX}/lib` altogether.

Reply:
> move us closer towards getting rid of any use of `-L#{HOMEBREW_PREFIX/}lib` and `-Wl,-rpath,#{HOMEBREW_PREFIX}/lib` altogether.

Yes, I'd really like to be able to do this too.

Reply:
> Whenever the host Glibc and host GCC are both not older, we should definitely not install the `glibc` formula.

This seems appropriate. Do you mean even if you do `brew install glibc`?

> > move us closer towards getting rid of any use of `-L#{HOMEBREW_PREFIX/}lib` and `-Wl,-rpath,#{HOMEBREW_PREFIX}/lib` altogether.
> 
> Yes, I'd really like to be able to do this too.

Same here 👍🏻

Reply:
Personal opinion is less people installing glibc the better, particularly when it comes to non-x86_64 platforms.

Happy to discuss about this though. Though it'll be a separate change.

Given I've not seen any objections, I'd like to merge this and https://github.com/Homebrew/homebrew-core/pull/109998 as soon as today (after some testing) unless anyone is against that.

Reply:
> Personal opinion is less people installing glibc the better, particularly when it comes to non-x86_64 platforms.

We don't support these. Let's not optimise for them at all until we have e.g. a means of generating portable Ruby there in CI.



> Given I've not seen any objections, I'd like to merge this and [Homebrew/homebrew-core#109998](https://github.com/Homebrew/homebrew-core/pull/109998) as soon as today (after some testing) unless anyone is against that.

👍🏻 although I think `binutils` is as much if not more of a problem here.

Reply:
> 👍🏻 although I think `binutils` is as much if not more of a problem here.

I'm not even sure why `gcc@11` depends on `binutils`, tbh

Reply:
> I'm not even sure why gcc@11 depends on binutils, tbh

@danielnachun Any ideas?

Reply:
Possibly it's just there to make sure we also install brewed `binutils` when we install brewed `gcc`. If that's the case it could be changed to a global dependency.  But we should make sure it's not needed for some other reason.

Reply:
> Possibly it's just there to make sure we also install brewed `binutils` when we install brewed `gcc`.

Why would we want to make sure of that?